java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll

```diff
@@ -0,0 +1,277 @@
+    // allow arbitrary field reads to balance out the implicit reads added above
+    exists(FieldRead fr | node1 = DataFlow::getFieldQualifier(fr) and node2.asExpr() = fr)
+  }
+}
+
+/**
+ * A configuration that a `MessageDigest` object is salted; i.e. it is used more than once
+ * or used with a non password.
+ */
+private class MessageDigestSaltedConfig extends DataFlow2::Configuration {
+  MessageDigestSaltedConfig() { this = "MessageDigestSaltedConfig" }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    source.asExpr() instanceof MessageDigestConstructor and
+    state instanceof HashZero
+  }
+
+  override predicate isSink(DataFlow::Node sink, DataFlow::FlowState state) {
+    messageDigestSink(sink, _, _) and
+    state instanceof HashSalted
+    or
+    messageDigestSink(sink, true, true) and
+    state instanceof HashOne
+  }
+
+  override predicate isAdditionalFlowStep(
+    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
+    DataFlow::FlowState state2
+  ) {
+    messageDigestFlowStep(node1, node2, true) and
+    state1 instanceof HashZero and
+    state2 instanceof HashOne
+    or
+    messageDigestFlowStep(node1, node2, true) and
+    state1 instanceof HashOne and
+    state2 instanceof HashSalted
+    or
+    messageDigestFlowStep(node1, node2, false) and
+    (state1 instanceof HashZero or state1 instanceof HashOne) and
+    state2 instanceof HashSalted
+  }
+
+  override predicate allowImplicitRead(DataFlow::Node node, DataFlow::Content c) {
+    // allow arbitrary implicit field read steps to allow wrapper methods that may go through fields to work
+    this.isAdditionalFlowStep(node, _, _, _) and
+    c instanceof DataFlow::FieldContent
+  }
+
+  override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+    // allow flow through arbitrary field reads to balance out the implicit reads added above
+    exists(FieldRead fr | node1 = DataFlow::getFieldQualifier(fr) and node2.asExpr() = fr)
+  }
+}
+
+/** Holds if there is loal dataflow between `node1` and `node2` in either direction. */
+private predicate localFlowBetween(DataFlow::Node node1, DataFlow::Node node2) {
+  DataFlow::localFlow(node1, node2) or DataFlow::localFlow(node2, node1)
+}
+
+/** Holds if `node` is a midpoint on a flow path from a source to a sink of `conf`. */
+private predicate flowPathMidPoint(DataFlow2::Configuration conf, DataFlow::Node node) {
+  exists(DataFlow2::PathNode source, DataFlow2::PathNode sink, DataFlow2::PathNode mid |
+    conf.hasFlowPath(source, sink) and
+    source.getASuccessor*() = mid and
+    mid.getASuccessor*() = sink and
+    node = mid.getNode()
+  )
+}
+
+/** Holds if `node` is on a path of a MessageDigest object that is used exactly once. */
+private predicate messageDigestFlowPath(DataFlow::Node node) {
+  flowPathMidPoint(any(MessageDigestUsedOnceConfig c), node) and
```

Comment:
Could the two configs be combined, always counting up to HashSalted but only sinking in the HashOne state? Can you give an example that requires two configs and these supporting predicates?

Reply:
I have tried that; however the use-use flow that *doesn't* change the flow state is still there; e.g. if you have 
```
md.update(password);
md.update(salt);
md.digest():
```
then there is still flow at state `HashOne` to the sink, as well as the flow at state `HashSalted`.

Reply:
Interesting, have you asked Tom if there's a good workaround for this?